gluten-core/src/main/scala/io/glutenproject/extension/PlanFallbackPolicy.scala

```diff
@@ -0,0 +1,164 @@
+
+import io.glutenproject.GlutenConfig
+import io.glutenproject.execution.BroadcastHashJoinExecTransformer
+import io.glutenproject.extension.columnar.TransformHints
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.{ColumnarBroadcastExchangeExec, ColumnarToRowExec, CommandResultExec, LeafExecNode, SparkPlan}
+import org.apache.spark.sql.execution.adaptive.{AdaptiveSparkPlanExec, BroadcastQueryStageExec, QueryStageExec, ShuffleQueryStageExec}
+import org.apache.spark.sql.execution.command.ExecutedCommandExec
+import org.apache.spark.sql.execution.exchange.Exchange
+
+// spotless:off
+/**
+ * Note, this rule should only fallback to row-based plan if there is no harm.
+ * The follow case should be handled carefully
+ *
+ * 1. A BHJ and the previous broadcast exchange is columnar
+ *    We should still make the BHJ columnar, otherwise it will fail if
+ *    the vanilla BHJ accept a columnar broadcast exchange, e.g.,
+ *
+ *    Scan                Scan
+ *      \                  |
+ *        \     Columnar Broadcast Exchange
+ *          \       /
+ *             BHJ
+ *              |
+ *       VeloxColumnarToRow
+ *              |
+ *           Project (unsupport columnar)
+ *
+ * 2. The previous shuffle exchange stage is a columnar shuffle exchange
+ *    We should use VeloxColumnarToRow rather than vanilla Spark ColumnarToRowExec, e.g.,
+ *
+ *             Scan
+ *              |
+ *    Columnar Shuffle Exchange
+ *              |
+ *       VeloxColumnarToRow
+ *              |
+ *           Project (unsupport columnar)
+ *
+ * @param isAdaptiveContext If is inside AQE
+ * @param originalPlan The vanilla SparkPlan without apply gluten transform rules
+ */
+// spotless:on
+case class PlanFallbackPolicy(isAdaptiveContext: Boolean, originalPlan: SparkPlan)
+  extends Rule[SparkPlan] {
+
+  private def countFallbacks(plan: SparkPlan): Int = {
+    var fallbacks = 0
+    def countFallback(plan: SparkPlan): Unit = {
+      plan match {
+        case _: QueryStageExec => // Another stage.
+        case _: CommandResultExec | _: ExecutedCommandExec => // ignore
+        // we plan exchange to columnar exchange in columnar rules and the exchange does not
+        // support columnar, so the output columnar is always false in AQE postStageCreationRules
+        case ColumnarToRowExec(s: Exchange) if s.supportsColumnar && isAdaptiveContext =>
+          countFallback(s)
+        case ColumnarToRowExec(p: GlutenPlan) =>
+          logDebug(s"Find a columnar to row for gluten plan:\n$p")
+          fallbacks = fallbacks + 1
+          countFallback(p)
+        case leafPlan: LeafExecNode if !leafPlan.isInstanceOf[GlutenPlan] =>
+          // Possible fallback for leaf node.
+          fallbacks = fallbacks + 1
+        case p => p.children.foreach(countFallback)
```

Comment:
Would you like to explain a bit more about this case? Thanks.

Reply:
sure, it is used to find the next cr2. Say a plan:
```
   Project
      |
RowToColumnar
      |
    Filter (unsupport columnar)
      |
ColumnarToRow
      |
     Scan

```

Reply:
Please ignore, I was refering to the previous case branch, not this one. However I noticed that was for leaf nodes like scans. Looking good to me.